src/cache/cache_las.c

```diff
@@ -859,17 +866,59 @@ __wt_las_sweep(WT_SESSION_IMPL *session)
 		 * should another thread remove the record before we do (not
 		 * expected for dropped trees), and the cursor remains
 		 * positioned in that case.
-		 *
+		 */
+		if (las_id >= cache->las_sweep_dropmin &&
+		    las_id <= cache->las_sweep_dropmax &&
+		    __bit_test(cache->las_sweep_dropmap,
+		    las_id - cache->las_sweep_dropmin)) {
+			WT_ERR(cursor->remove(cursor));
+			++remove_cnt;
+			continue;
+		}
+
+		/*
 		 * TODO it would also be good to remove entries in lookaside
 		 * from live files that have aged out.  If we track for each
 		 * entry whether it was the on-page value chosen by
 		 * reconciliation, we can safely remove entries from that point
 		 * on (for the given key) that are visible to all readers.
 		 */
-		if (las_id >= cache->las_sweep_dropmin &&
-		    las_id <= cache->las_sweep_dropmax &&
-		    __bit_test(cache->las_sweep_dropmap,
-		    las_id - cache->las_sweep_dropmin)) {
+		WT_ERR(cursor->get_value(cursor,
+		    &txnid, &las_timestamp, &upd_type, &las_value));
+		val_ts = NULL;
+#ifdef HAVE_TIMESTAMPS
+		WT_ASSERT(session, las_timestamp.size == WT_TIMESTAMP_SIZE);
+		memcpy(&timestamp, las_timestamp.data, las_timestamp.size);
+		val_ts = &timestamp;
+#endif
+		/*
+		 * If this entry isn't globally visible we cannot remove it.
+		 * If it is visible then perform additional checks to see
+		 * whether it has aged out of a live file.
+		 */
+		if (__wt_txn_visible_all(session, txnid, val_ts)) {
+			/*
+			 * Save our key for comparing with older entries if we
+			 * don't have one or it is different.
+			 */
+			if (saved_key == NULL ||
+			    saved_key->size != las_key.size ||
+			    memcmp(saved_key->data,
+			    las_key.data, las_key.size) != 0) {
+				WT_ERR(__wt_cursor_get_raw_key(
+				    cursor, saved_key));
+				if (!WT_DATA_IN_ITEM(saved_key))
+					WT_ERR(__wt_buf_set(session, saved_key,
+					    saved_key->data, saved_key->size));
+				/*
+				 * If it is a birthmark record, even the
+				 * oldest key can be removed because it is
+				 * on disk already and now visible. Otherwise
+				 * we don't remove the first entry.
+				 */
+				if (upd_type != WT_UPDATE_BIRTHMARK)
+					continue;
+			}
```

Comment:
The `__wt_txn_visible_all` call is relatively expensive (including a timestamp read-lock). I think we could avoid repeating that call once we find a visible record with something like this:
```
bool found_visible = false;

/* Save a copy of each new key. */
if (saved_key == NULL || saved_key->size != las_key.size ||
    memcmp(saved_key->data, las_key.data, las_key.size) != 0) {
        found_visible = false;
        WT_ERR(__wt_cursor_get_raw_key(cursor, saved_key));
        if (!WT_DATA_IN_ITEM(saved_key))
                WT_ERR(__wt_buf_set(session, saved_key,
                    saved_key->data, saved_key->size));
}

/*
 * If this entry isn't globally visible we cannot remove it.
 * If it is visible then perform additional checks to see
 * whether it has aged out of a live file.
 */
if (!found_visible) {
        if (!__wt_txn_visible_all(session, txnid, val_ts))
                continue;
        found_visible = true;
         
        /*
         * If it is a birthmark record, even the oldest key
         * can be removed because it is on disk already and
         * now visible. Otherwise we don't remove the first
         * entry.
         */
        if (upd_type != WT_UPDATE_BIRTHMARK)
                continue;
}
WT_ERR(cursor->remove(cursor));
++remove_cnt;
```

Reply:
Hmm, I'm torn on this. Your suggested code is more convoluted than what I have. If it is important to avoid the visibility call for performance it would work. But the question is, do we care? This is the sweep thread, not a critical path. The readlock mentioned is only for odd-sized timestamps and is not used in 8-byte timestamps.

Reply:
@michaelcahill we're going to let you be the tiebreaker here! @keithbostic falls more on the side of avoiding the visibility call for performance. I fall on the side of keeping the flow more readable (IMO) because it isn't a perf path.